### PR TITLE
Achievement

### PR DESCRIPTION
**Pull request by forestik**

### CODE REVIEW

File: core/src/main/java/greencity/webcontroller/ManagementAchievementController.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+import org.springframework.validation.BindingResult;
+import org.springframework.web.bind.annotation.*;
+
+import javax.validation.Valid;
+import java.util.List;
+
+import static greencity.dto.genericresponse.GenericResponseDto.buildGenericResponseDto;
+
+@Controller
+@AllArgsConstructor
```
Comment (vterlyha) [inline core/src/main/java/greencity/webcontroller/ManagementAchievementController.java:28]: Can be changed to @Requiredargscontructor

File: core/src/main/java/greencity/webcontroller/ManagementAchievementController.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    /**
+     * Method which deteles {@link AchievementVO} by given id.
+     *
+     * @param id of {@link AchievementVO}
+     * @return {@link ResponseEntity}
+     * @author Orest Mamchuk
+     */
+    @DeleteMapping("delete/{id}")
```
Comment (vterlyha) [inline core/src/main/java/greencity/webcontroller/ManagementAchievementController.java:80]: 'delete' relative path not needed. Please, remove it.

File: core/src/main/java/greencity/webcontroller/ManagementAchievementController.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        return ResponseEntity.status(HttpStatus.OK).body(listId);
+    }
+
+    /**
+     * Method for getting econews by id.
+     *
+     * @param id of Eco New
+     * @return {@link AchievementVO} instance.
+     */
+    @GetMapping("/find")
```
Comment (vterlyha) [inline core/src/main/java/greencity/webcontroller/ManagementAchievementController.java:105]: @RequestParam("id") Long id - move this to path parama as {id}, and remove 'find' from relative path

File: core/src/main/java/greencity/webcontroller/ManagementAchievementController.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        return ResponseEntity.status(HttpStatus.OK).body(achievementService.findById(id));
+    }
+
+    /**
+     * Method which updates {@link AchievementVO}.
+     *
+     * @param achievementManagementDto of {@link AchievementManagementDto}.
+     * @return {@link GenericResponseDto} with of operation and errors fields.
+     */
+    @PutMapping("/")
```
Comment (vterlyha) [inline core/src/main/java/greencity/webcontroller/ManagementAchievementController.java:116]: / - unecessary

File: dao/src/main/java/greencity/entity/User.java [hunk 5]
```diff
@@ -106,6 +106,9 @@
         inverseJoinColumns = @JoinColumn(name = "friend_id", referencedColumnName = "id"))
     private List<User> userFriends = new ArrayList<>();
 
+    @OneToMany(mappedBy = "user", fetch = FetchType.EAGER, cascade = CascadeType.ALL)
```
Comment (vterlyha) [inline dao/src/main/java/greencity/entity/User.java:109]: FetchType.EAGER - should be LAZY and get it in query only if it is needed

File: dao/src/main/java/greencity/repository/AchievementRepo.java [hunk 6]
```diff
@@ -1,9 +1,30 @@
...
 public interface AchievementRepo extends JpaRepository<Achievement, Long> {
+    /**
+     * Method returns {@link EcoNews} by search query and page.
+     *
+     * @param paging {@link Pageable}.
+     * @param query  query to search.
+     * @return list of {@link EcoNews}.
+     * @author
+     */
+    @Query("SELECT DISTINCT a FROM Achievement a "
```
Comment (vterlyha) [inline dao/src/main/java/greencity/repository/AchievementRepo.java:21]: Bad approach

File: service-api/src/main/java/greencity/dto/achievement/AchievementManagementDto.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+import greencity.dto.achievementcategory.AchievementCategoryDto;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import javax.validation.constraints.Min;
+import javax.validation.constraints.NotNull;
+import java.util.List;
+
+@Data
```
Comment (vterlyha) [inline service-api/src/main/java/greencity/dto/achievement/AchievementManagementDto.java:12]: Do you need Data annotation here ?

File: service-api/src/main/java/greencity/dto/achievement/AchievementPostDto.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
+package greencity.dto.achievement;
+
+import greencity.dto.achievementcategory.AchievementCategoryDto;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.util.List;
+
+@Data
```
Comment (vterlyha) [inline service-api/src/main/java/greencity/dto/achievement/AchievementPostDto.java:10]: Same here

File: service-api/src/main/java/greencity/dto/achievement/AchievementPostDto.java [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+@NoArgsConstructor
+@AllArgsConstructor
+public class AchievementPostDto {
```
Comment (vterlyha) [inline service-api/src/main/java/greencity/dto/achievement/AchievementPostDto.java:13]: AchievementPostDto  - looks a lot of duplication from AchievementManagementDto. Think of inheritance to composition

File: service/src/main/java/greencity/mapping/UserVOMapper.java [hunk 10]
```diff
@@ -36,6 +38,23 @@ protected UserVO convert(User user) {
                     .build())
                 .collect(Collectors.toList());
         }
+
+        List<UserAchievementVO> userAchievements = new ArrayList<>();
+        if (user.getUserAchievements() != null) {
+            userAchievements = user.getUserAchievements()
+                .stream().map(userAchievement -> UserAchievementVO.builder()
```
Comment (vterlyha) [inline service/src/main/java/greencity/mapping/UserVOMapper.java:45]: Move mapping operation to converter. And for VerifyEmailVO  and  for userFriends  also

File: service/src/main/java/greencity/mapping/UserVOMapper.java [hunk 11]
```diff
@@ -69,6 +88,7 @@ protected UserVO convert(User user) {
             .showEcoPlace(user.getShowEcoPlace())
             .showLocation(user.getShowLocation())
             .lastActivityTime(user.getLastActivityTime())
+            .userAchievements(userAchievements)
```
Comment (vterlyha) [inline service/src/main/java/greencity/mapping/UserVOMapper.java:91]: Move everything to converter

File: service/src/main/java/greencity/service/AchievementServiceImpl.java [hunk 12]
```diff
@@ -1,22 +1,61 @@
...
+     * @author Orest Mamchuk
+     */
+    @Override
+    public AchievementVO save(AchievementPostDto achievementPostDto) {
+        Achievement achievement = modelMapper.map(achievementPostDto, Achievement.class);
+        AchievementCategoryVO achievementCategoryVO =
+            achievementCategoryService.findByName(achievementPostDto.getAchievementCategory().getName());
+        achievement.getTranslations().forEach(adviceTranslation -> adviceTranslation.setAchievement(achievement));
+        achievement.setAchievementCategory(modelMapper.map(achievementCategoryVO, AchievementCategory.class));
+        AchievementVO map = modelMapper.map(achievementRepo.save(achievement), AchievementVO.class);
```
Comment (vterlyha) [inline service/src/main/java/greencity/service/AchievementServiceImpl.java:48]: map ?

File: service/src/main/java/greencity/service/AchievementServiceImpl.java [hunk 13]
```diff
@@ -31,4 +70,116 @@
...
+     * @author Orest Mamchuk
+     */
+    @Override
+    public PageableAdvancedDto<AchievementVO> findAll(Pageable page) {
+        Page<Achievement> pages = achievementRepo.findAll(page);
+        List<AchievementVO> achievementVOS = pages
+            .stream()
+            .map(achievement -> modelMapper.map(achievement, AchievementVO.class))
+            .collect(Collectors.toList());
+        return new PageableAdvancedDto<>(
```
Comment (vterlyha) [inline service/src/main/java/greencity/service/AchievementServiceImpl.java:86]: Move creation to private method

File: service/src/main/java/greencity/service/AchievementServiceImpl.java [hunk 14]
```diff
@@ -31,4 +70,116 @@
...
+     * @author Orest Mamchuk
+     */
+    @Override
+    public PageableAdvancedDto<AchievementVO> searchAchievementBy(Pageable paging, String query) {
+        Page<Achievement> page = achievementRepo.searchAchievementsBy(paging, query);
+
+        List<AchievementVO> achievementVOS = page.stream()
+            .map(achievement -> modelMapper.map(achievement, AchievementVO.class))
+            .collect(Collectors.toList());
+        return new PageableAdvancedDto<>(
```
Comment (vterlyha) [inline service/src/main/java/greencity/service/AchievementServiceImpl.java:110]: Same here

File: service/src/main/java/greencity/service/AchievementServiceImpl.java [hunk 15]
```diff
@@ -31,4 +70,116 @@
...
+        achievement.getTranslations()
+            .forEach(achievementTranslation -> {
+                AchievementTranslationVO achievementTranslationVO = achievementManagementDto
+                    .getTranslations().stream()
+                    .filter(newTranslation -> newTranslation.getLanguage().getCode()
+                        .equals(achievementTranslation.getLanguage().getCode()))
+                    .findFirst().get();
+                achievementTranslation.setTitle(achievementTranslationVO.getTitle());
+                achievementTranslation.setDescription(achievementTranslationVO.getDescription());
+                achievementTranslation.setMessage(achievementTranslationVO.getMessage());
```
Comment (vterlyha) [inline service/src/main/java/greencity/service/AchievementServiceImpl.java:179]: Move to private method

**Review by Yura09 — APPROVED** [review]

File: core/src/main/java/greencity/webcontroller/ManagementAchievementController.java (hunk 1 above)
Comment (forestik) [reply core/src/main/java/greencity/webcontroller/ManagementAchievementController.java]: Done

File: core/src/main/java/greencity/webcontroller/ManagementAchievementController.java (hunk 2 above)
Comment (forestik) [reply core/src/main/java/greencity/webcontroller/ManagementAchievementController.java]: Done

File: core/src/main/java/greencity/webcontroller/ManagementAchievementController.java (hunk 3 above)
Comment (forestik) [reply core/src/main/java/greencity/webcontroller/ManagementAchievementController.java]: Done

File: core/src/main/java/greencity/webcontroller/ManagementAchievementController.java (hunk 4 above)
Comment (forestik) [reply core/src/main/java/greencity/webcontroller/ManagementAchievementController.java]: Done

File: dao/src/main/java/greencity/entity/User.java (hunk 5 above)
Comment (forestik) [reply dao/src/main/java/greencity/entity/User.java]: Done

File: service/src/main/java/greencity/mapping/UserVOMapper.java (hunk 11 above)
Comment (forestik) [reply service/src/main/java/greencity/mapping/UserVOMapper.java]: Done

File: service-api/src/main/java/greencity/dto/achievement/AchievementManagementDto.java (hunk 7 above)
Comment (forestik) [reply service-api/src/main/java/greencity/dto/achievement/AchievementManagementDto.java]: Done

File: service-api/src/main/java/greencity/dto/achievement/AchievementPostDto.java (hunk 8 above)
Comment (forestik) [reply service-api/src/main/java/greencity/dto/achievement/AchievementPostDto.java]: Done

File: service-api/src/main/java/greencity/dto/achievement/AchievementPostDto.java (hunk 9 above)
Comment (forestik) [reply service-api/src/main/java/greencity/dto/achievement/AchievementPostDto.java:15]: Done

File: service/src/main/java/greencity/mapping/UserVOMapper.java (hunk 10 above)
Comment (forestik) [reply service/src/main/java/greencity/mapping/UserVOMapper.java]: Done

File: service/src/main/java/greencity/service/AchievementServiceImpl.java (hunk 12 above)
Comment (forestik) [reply service/src/main/java/greencity/service/AchievementServiceImpl.java]: Done

File: service/src/main/java/greencity/service/AchievementServiceImpl.java (hunk 13 above)
Comment (forestik) [reply service/src/main/java/greencity/service/AchievementServiceImpl.java:91]: Done

File: service/src/main/java/greencity/service/AchievementServiceImpl.java (hunk 14 above)
Comment (forestik) [reply service/src/main/java/greencity/service/AchievementServiceImpl.java]: Done

File: service/src/main/java/greencity/service/AchievementServiceImpl.java (hunk 15 above)
Comment (forestik) [reply service/src/main/java/greencity/service/AchievementServiceImpl.java:182]: Done

**Review by vterlyha — APPROVED** [review]

**Review by Markiyan000 — APPROVED** [review]

**Pull request merged**